### ISSUE TITLE
AutoComplete/Template Variables: Prometheus auto-complete and template variables can be very slow to load on large clusters

### ISSUE THREAD

**Issue by harry671003** [grafana/grafana#87871]
### What happened?

When a Prometheus compatible data source such as Cortex is used with millions of time series, the Grafana auto-complete and template variables can be very slow to load.

https://github.com/grafana/grafana/assets/842522/57eefbe4-4afa-47a7-a218-00e3244ed005

### What did you expect to happen?

Auto complete is expected to be much quicker to be useful to the users of Grafana.

### How do we reproduce it?

Video recording attached.

### Environment (with versions)?

Grafana: Amazon Managed Grafana
OS: Mac OS X
Browser: Firefox

### Grafana platform?

Other

### Datasource(s)?

Prometheus

**Comment by harry671003** [comment]
Recently Prometheus has introduced a new LIMIT parameter for the metadata APIs. 
https://github.com/prometheus/prometheus/pull/13396

Also see the HTTP reference:
- https://prometheus.io/docs/prometheus/latest/querying/api/#finding-series-by-label-matchers
- https://prometheus.io/docs/prometheus/latest/querying/api/#getting-label-names
- https://prometheus.io/docs/prometheus/latest/querying/api/#querying-label-values


Would it be a good idea to pass a configurable limit (eg: 10K) when calling these APIs in auto-complete?

**Comment by harry671003** [comment]
Duplicate of other issue?
- https://github.com/grafana/grafana/issues/85444
- https://github.com/grafana/grafana/issues/84612

**Comment by bohandley** [comment]
Yes, this is a known issue for customers that have millions of metrics. Adding the limit for metrics sounds sensible. Although, we worry about autocomplete not having all metrics and users not having the full autocomplete capability.

**Comment by a2intl** [comment]
Hi @bohandley , just wanted to give this an issue upvote, we're running into slowness during autocomplete developing on our mimir cluster, loading about 6.2k metric names and 1.6k label names from mimir. We're seeing the performance on `/metadata` and `/labels` on freshly-started grafana pods be fine, and then slowly degrade over time, leading to autocomplete becoming slow & unresponsive, so part of the fix might be better server-cache management.
